lightly/api/download.py

```diff
@@ -0,0 +1,186 @@
+import concurrent.futures
+import os
+import pathlib
+import shutil
+import threading
+import warnings
+from concurrent.futures import ThreadPoolExecutor
+from typing import List
+
+import cv2
+import requests
+import tqdm
+from lightly.api import utils
+from PIL import Image
+
+
+def cv2_to_pil(image):
+    """Convert cv2 image to PIL.Image"""
+    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+    return Image.fromarray(image)
+
+
+def download_image(url: str, session: requests.Session = None):
+    """Downloads an image from an url
+
+    Args:
+        url (str):
+            Url where image is stored
+        session (requests.Session):
+            Make request using the given session
+
+    Returns:
+        PIL.Image
+    """
+    req = requests if session is None else session
+    response = req.get(url, stream=True)
+    return Image.open(response.raw)
+
+
+def download_all_video_frames(url: str, as_pil_image: int = True):
+    """Lazily retrieves all frames from a video
+
+    Args:
+        url (str):
+            Url where video is stored
+        as_pil_image (bool):
+            Whether to return the frame as PIL.Image
+
+    Returns:
+        generator:
+            Loads and returns a single frame per step.
+            Frames are of type PIL.Image if `as_pil_image` is `True`,
+            otherwise a np.array is returned
+
+    """
+    video = cv2.VideoCapture(url)
```

Comment:
Yes I am aware. cv2 is the only library that I am aware of that allows this workflow. I would love to avoid it if possible. Maybe it would make more sense to download the video in the background to disk and then use the videoreader from torchvision. This would be more consistent with the non-iterable dataset.